internal/jobcontainers/path.go

```diff
@@ -0,0 +1,143 @@
+package jobcontainers
+
+import (
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/Microsoft/hcsshim/internal/winapi"
+	"github.com/pkg/errors"
+	"golang.org/x/sys/windows"
+)
+
+func searchPathForExe(fileName, pathsToSearch string) (string, error) {
+	fileNamePtr, err := windows.UTF16PtrFromString(fileName)
+	if err != nil {
+		return "", err
+	}
+
+	pathsToSearchPtr, err := windows.UTF16PtrFromString(pathsToSearch)
+	if err != nil {
+		return "", err
+	}
+
+	extension, err := windows.UTF16PtrFromString(".exe")
+	if err != nil {
+		return "", err
+	}
+
+	path := make([]uint16, windows.MAX_PATH)
+	_, err = winapi.SearchPath(
+		pathsToSearchPtr,
+		fileNamePtr,
+		extension,
+		windows.MAX_PATH,
+		&path[0],
+		nil,
+	)
+	if err != nil {
+		return "", err
+	}
+
+	exePath := windows.UTF16PtrToString(&path[0])
+	// Need to check if we just found a directory with the name of the executable and
+	// .exe at the end. ping.exe is a perfectly valid directory name for example.
+	attrs, err := os.Stat(exePath)
+	if err != nil {
+		return "", &os.PathError{Path: exePath, Op: "stat", Err: err}
+	}
+
+	if attrs.IsDir() {
+		return "", fmt.Errorf("found directory instead of executable %q", exePath)
+	}
+
+	return exePath, nil
+}
+
+// Returns the system paths (system32, system, and windows) as a search path,
+// including a terminating ;.
+//
+// Typical output would be `C:\WINDOWS\system32;C:\WINDOWS\System;C:\WINDOWS;`
+func getSystemPaths() (string, error) {
+	var searchPath string
+	systemDir, err := windows.GetSystemDirectory()
+	if err != nil {
+		return "", errors.Wrap(err, "failed to get system directory")
+	}
+	searchPath += systemDir + ";"
+
+	windowsDir, err := windows.GetWindowsDirectory()
+	if err != nil {
+		return "", errors.Wrap(err, "failed to get Windows directory")
+	}
+
+	searchPath += windowsDir + "\\System;" + windowsDir + ";"
+	return searchPath, nil
+}
+
+// Gets the application name from the commandline provided and returns the commandline
+// as is or quoted if found necessary.
+func getApplicationName(commandLine, workingDirectory, pathEnv string) (string, string, error) {
+	var (
+		searchPath string
+		result     string
+	)
+	systemPaths, err := getSystemPaths()
+	if err != nil {
+		return "", "", err
+	}
+
+	if workingDirectory != "" {
+		searchPath += workingDirectory + ";"
+	}
+	searchPath += systemPaths + ";"
+	searchPath += pathEnv
+
+	if searchPath[len(searchPath)-1] == ';' {
+		searchPath = searchPath[:len(searchPath)-1]
+	}
+
+	// Application name was quoted, just search directly.
+	if commandLine != "" && commandLine[0] == '"' {
+		index := strings.Index(commandLine[1:], "\"")
+		path, err := searchPathForExe(commandLine[1:index+1], searchPath)
```

Comment:
This will panic if `strings.Index` returns -1. I also don't quite understand the purpose of this section. Why only search for the first directory in the exe's path?

Reply:
We're just searching using the quoted contents

Reply:
e.g. for `"hello goodbye.exe" -lots -of -flags`, we'd throw hello goodbye.exe into the search

Reply:
But I'll need to check for -1 if there's only one quote good catch

Reply:
is there a guarantee that the commandline has a closing `"`?